onyo/lib/items.py

```diff
@@ -0,0 +1,163 @@
+
+
+class Item(DotNotationWrapper):
+    """An Item an Inventory can potentially track.
+
+    The main purpose of this class is to attach pseudo-keys
+    and alias resolution to things that can be inventoried.
+    That's directories and YAML-files.
+
+    The initializer methods are referenced in the `PSEUDO_KEYS`
+    mapping. They are called, when a pseudo-key is first
+    accessed (see `self.__getitem__`). This allows to distinguish
+    a meaningful `None` (<unset>) from a not yet evaluated
+    pseudo-key.
+    """
+
+    def __init__(self,
+                 item: Mapping[_KT, _VT] | Path | None = None,
+                 repo: onyo.lib.onyo.OnyoRepo | None = None,
+                 **kwargs: _VT):
+        super().__init__()
+        self.repo: onyo.lib.onyo.OnyoRepo | None = repo
+        self.update(onyo.lib.pseudokeys.PSEUDO_KEYS)
+        self._path = None
+
+        if isinstance(item, Path):
+            assert item.is_absolute()  # currently no support for relative. This is how all existing code should work ATM.
+            self.update_from_path(item)
+        elif item is not None:
+            self.update(item)
+
+        if kwargs:
+            self.update(**kwargs)
+
+    def __setitem__(self, key, value):
+        key = resolve_alias(key)
+        super().__setitem__(key, value)
+
+    def __getitem__(self, key):
+        key = resolve_alias(key)
+        value = super().__getitem__(key)
+        if key in onyo.lib.pseudokeys.PSEUDO_KEYS and \
+                isinstance(value, onyo.lib.pseudokeys.PseudoKey):
+            # Value still is the pseudo-key definition.
+            # Actually load and set the response as the new value.
+            new_value = value.implementation(self)
+            self[key] = new_value
+            return new_value
+        return value
+
+    def __delitem__(self, key):
+        return super().__delitem__(resolve_alias(key))
+
+    def __contains__(self, key):
+        return super().__contains__(resolve_alias(key))
+
+    def get(self, key, default=None):
+        return super().get(resolve_alias(key), default=default)
+
+    def update_from_path(self, path: Path):
+        """Update internal dictionary from a YAML file."""
+        # TODO: Potentially account for being pointed to
+        #       a directory or an .onyoignore'd file.
+        from onyo.lib.utils import get_asset_content
+        self._path = path
+        if self['onyo.is.asset']:
+            loader = self.repo.get_asset_content if self.repo else get_asset_content
+            self.update(loader(path))
+
+    # TODO: git metadata:
+    # def fill_created(self, what: str | None):
+    #     # self.repo.git.xxx   (self.get('onyo.path.file'))
+    #     # fill in all created keys
+    #     # switch `what` -> return whatever part was requesting this
+    #     pass#raise NotImplementedError
+    #
+    # def fill_modified(self, what: str | None):
+    #     pass#raise NotImplementedError
+
+    def get_path_absolute(self):
```

Comment:
Does not need to be addressed in this PR, but I think I'd prefer these to all be private methods.

They're used internally to populate the pseudokeys, but everything else uses the pseudokey namespace.